kube-client/src/api/entry.rs

```diff
@@ -0,0 +1,394 @@
+//! API helpers for get-or-create and get-and-modify patterns
+//!
+//! [`Api::entry`] is the primary entry point for this API.
+
+// Import used in docs
+#[allow(unused_imports)] use std::collections::HashMap;
+use std::fmt::Debug;
+
+use kube_core::{params::PostParams, Resource};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{Api, Result};
+
+impl<K: Resource + Clone + DeserializeOwned + Debug> Api<K> {
+    /// Gets a given object's "slot" on the Kubernetes API, designed for "get-or-create" and "get-and-modify" patterns
+    ///
+    /// This is similar to [`HashMap::entry`], but [`Entry`] must be [`Entry::sync`]ed for changes to be persisted.
+    pub async fn entry<'a>(&'a self, name: &'a str) -> Result<Entry<'a, K>> {
+        Ok(match self.get_opt(name).await? {
+            Some(object) => Entry::Occupied(OccupiedEntry {
+                api: self,
+                object,
+                dirtiness: Dirtiness::Clean,
+            }),
+            None => Entry::Vacant(VacantEntry { api: self, name }),
+        })
+    }
+}
+
+#[derive(Debug)]
+/// A view into a single object, with enough context to create or update it
+///
+/// See [`Api::entry`].
+pub enum Entry<'a, K> {
+    /// An object that either exists on the server, or has been created locally (and is awaiting synchronization)
+    Occupied(OccupiedEntry<'a, K>),
+    /// An object that does not exist
+    Vacant(VacantEntry<'a, K>),
+}
+
+impl<'a, K> Entry<'a, K> {
+    /// Borrow the object, if it exists (on the API, or queued for creation using [`Entry::or_insert`])
+    pub fn get(&self) -> Option<&K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    /// Borrow the object mutably, if it exists (on the API, or queued for creation using [`Entry::or_insert`])
+    ///
+    /// [`Entry::sync`] must be called afterwards for any changes to be persisted.
+    pub fn get_mut(&mut self) -> Option<&mut K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get_mut()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    /// Let `f` modify the object, if it exists (on the API, or queued for creation using [`Entry::or_insert`])
+    ///
+    /// [`Entry::sync`] must be called afterwards for any changes to be persisted.
+    pub fn and_modify(self, f: impl FnOnce(&mut K)) -> Self {
```

Comment:
could/should also have smaller `no_run` doc examples for these other main interfaces (or_insert, get_mut, sync)

`Api::entry` is the main thing that needs one though

Reply:
I added an example to `Api::entry` that also covers `or_insert`, `and_modify`, and `sync`.